### ISSUE TITLE
Add support for bandwidth limit

### ISSUE THREAD

**Issue by rubenrua** [thecoshman/http#102]
The missing feature I need for this incredible software.

```
http --limit 100kbps
```

**Comment by thecoshman** [comment]
@rubenrua do you need/want asymmetric limiting for this? What about limiting per client and/or overall?

We'll look into if the library we use for the underlying http can offer this, if so, we'll expose controls for it.

**Comment by nabijaczleweli** [comment]
Neither `iron` (the library we use directly), nor `hyper` (the library `iron` uses) exposes this.

On POSIX, [trickle](https://github.com/mariusae/trickle) *would* be the best and simplest option, but it doesn't work (apparently because it doesn't work on forking processes, and `iron` does that gratuitously); your next-best bet would be to use something like wondershaper or your firewall (nftables has this, SO says iptables does too).
Unfortunately, I found zero non-root options, though if you do find one that works, I'd be delighted to hear about it.

On Win32, between [this](https://superuser.com/questions/135719/way-to-limit-bandwidth-of-programs-on-windows) SU answer and [this](https://www.howtogeek.com/347711/how-to-limit-any-applications-bandwidth-on-windows) article, there seems to be just one free option – [TMeter](www.tmeter.ru/en/).

I'll open an issue upstream but I wouldn't hold my breath.

**Comment by nabijaczleweli** [comment]
Tracking upstream issue https://github.com/hyperium/hyper/issues/2163

**Comment by nabijaczleweli** [comment]
So I had a think about this issue, and:
  * there could theoretically be a middleware that'd replace the body with a wrapper that has a `Read` impl that sleeps (ideologically revolting, but whatever), but:
    * in the simple case, it'd be per-request, essentially saturating the band anyway with many enough requests,
    * in the more complex case, each request body would have to synchronise against some global lock? queue? and cooperate with every other request body. This would work, but would also be a nightmare to implement correctly and test.
  * in-application bandwidth limiting is (at this current moment) a pain the arse, both to implement and manage; if I had to deploy a limit, I'd manage it centrally at the firewall.

For this sort of thing an ideal solution would be `trickle`, which is fully user-space and non-root, but it doesn't work to control forking processes (I'm assuming for the same synchronisation reason).

A per-request bandwidth limit is possible, but I'm not sure if that's what @rubenrua had in mind, or if it'd work in his use-case anyway.

**Comment by thecoshman** [comment]
Annoyingly, I can see it being useful to be able to limit both the overall servers bandwidth usage and the per request/client bandwidth. 

One possible way to limit the overall servers bandwidth might be to run `http` within a container, fairly sure you can do this.

**Comment by rubenrua** [comment]
I use network throttling in the browser (firefox or chrome) to do test related with multimedia. Also when a need to configure the throttling in the server for a test outside the browser I have to build NGINX with `ngx_http_limit_req_module`.

I really like your sw, and as I said before I mess these feature.

> A per-request bandwidth limit is possible, but I'm not sure if that's what @rubenrua had in mind, or if it'd work in his use-case anyway.

Yes, For me per-request bw limit is enough.

**Comment by nabijaczleweli** [comment]
Alright, @rubenrua can you try the `feature/102-request-bandwidth` branch and `--request-bandwith` option?

**Comment by rubenrua** [comment]
Sorry for the late response.

Perfect

```
./target/release/http --request-bandwidth 10 ~/tmp &
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0 7260k    0 26257    0     0    860      0  2:24:05  0:00:30  2:23:35   862^C

./target/release/http --request-bandwidth 100 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0 7260k    0 11533    0     0    851      0  2:25:36  0:00:13  2:25:23   851^C 

./target/release/http --request-bandwidth 1000 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0 7260k    0 12402    0     0    855      0  2:24:56  0:00:14  2:24:42   853^C

./target/release/http --request-bandwidth 10000 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  1 7260k    1 83290    0     0   8531      0  0:14:31  0:00:09  0:14:22  8552^C

./target/release/http --request-bandwidth 100000 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  9 7260k    9  686k    0     0  84829      0  0:01:27  0:00:08  0:01:19 84655^C

./target/release/http --request-bandwidth 1000000 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 49 7260k   49 3610k    0     0   780k      0  0:00:09  0:00:04  0:00:05  780k^C

./target/release/http --request-bandwidth 10000000 ~/tmp
curl http://localhost:8000/data -O
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 7260k  100 7260k    0     0  6600k      0  0:00:01  0:00:01 --:--:-- 6606k
```

I get bad results with very low bit-rates (request-bandwidth < 500) but for me it's not a problem.

Thanks.

**Comment by nabijaczleweli** [comment]
Great! Would you mind trying https://github.com/thecoshman/http/commit/419bfe9a21ee66faad574eb59b55c5ccc5044bb1 from the updated `feature/102-request-bandwidth` as well? I had it kicking around labelled "testing accuracy", so maybe it'll be a bit better on that front. Then I'll prepare a release.

**Comment by nabijaczleweli** [comment]
Released in [`v1.10.0`](https://github.com/thecoshman/http/releases/tag/v1.10.0), but feedback still appreciated
